File: ocr-service/scorecard_parser.py
# Copyright (c) 2026 RAJ.Y — All rights reserved.
# CH_Eleven Fantasy Cricket Platform
# https://github.com/RAJ-Y/ch-eleven
"""
Scorecard Parser - Extracts batting/bowling tables from OCR text
Handles cricket scorecards from PDFs, images, DOCX
"""

import re
import pandas as pd
from typing import Dict, List, Tuple, Any

class ScorecardParser:
    """Parse cricket scorecard text to extract batting and bowling stats"""

    def __init__(self):
        self.batting_keywords = ['batsman', 'batsmen', 'player', 'runs', 'balls', 'fours', 'sixes']
        self.bowling_keywords = ['bowler', 'overs', 'maidens', 'wickets', 'runs', 'economy']

    def parse_text(self, text: str, inning: int = 1) -> Dict[str, Any]:
        """
        Parse OCR extracted text into structured scorecard data
        
        Args:
            text: Raw OCR text from scorecard
            inning: 1 or 2 (match inning number)
            
        Returns:
            Dict with batting, bowling, and match info
        """
        result = {
            "inning": inning,
            "batting": [],
            "bowling": [],
            "match_info": {},
            "confidence": 0.0,
        }

        # Extract match details
        result["match_info"] = self._extract_match_info(text)
        
        # Extract batting table
        result["batting"] = self._extract_batting(text)
        
        # Extract bowling table
        result["bowling"] = self._extract_bowling(text)

        # Calculate confidence score
        result["confidence"] = self._calc_confidence(result)

        return result

    def _extract_match_info(self, text: str) -> Dict[str, str]:
        """Extract match details (teams, ground, date)"""
        info = {}
        
        # Try to find team names (usually capitalized phrases)
        lines = text.split('\n')
        for line in lines[:20]:  # Check first 20 lines
            if 'vs' in line.lower() or 'versus' in line.lower():
                info['match'] = line.strip()
        
        # Extract ground/venue
        for keyword in ['ground', 'venue', 'location', 'cricket ground']:
            pattern = rf'{keyword}.*?([A-Z][^,\n]+(?:,[^,\n]+)?)'
            match = re.search(pattern, text, re.IGNORECASE)
            if match:
                info['ground'] = match.group(1).strip()
                break

        # Extract date
        date_pattern = r'\d{1,2}[-/]\d{1,2}[-/]\d{2,4}'
        date_match = re.search(date_pattern, text)
        if date_match:
            info['date'] = date_match.group(0)

        return info

    def _extract_batting(self, text: str) -> List[Dict[str, Any]]:
        """Extract batting statistics from text"""
        batting_list = []
        
        # Split text into lines
        lines = [line.strip() for line in text.split('\n') if line.strip()]
        
        # Look for batting section
        batting_start = -1
        for i, line in enumerate(lines):
            if re.search(r'(?:batsman|batting|inning.*batting)', line, re.IGNORECASE):
                batting_start = i + 1
                break
        
        if batting_start == -1:
            # No clear section header, try to find batting patterns
            batting_start = 0

        # Parse individual batting rows
        for i, line in enumerate(lines[batting_start:]):
            # Stop at bowling section
            if re.search(r'(?:bowler|bowling)', line, re.IGNORECASE):
                break

            # Match batting line pattern: Name Runs Balls 4s 6s Status
            # Example: "Raj007 30 7 1 2 not out"
            batting_match = re.match(
                r'^([A-Za-z0-9\s\(\)]+?)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s*(.*?)$',
                line
            )

            if batting_match:
                batsman_name = batting_match.group(1).strip()
                
                # Skip if it's a header line
                if batsman_name.lower() in ['no', 'batsman', 'runs', 'balls']:
                    continue

                batting_data = {
                    "batsman": batsman_name,
                    "runs": int(batting_match.group(2)),
                    "balls": int(batting_match.group(3)),
                    "4s": int(batting_match.group(4)),
                    "6s": int(batting_match.group(5)),
                    "status": batting_match.group(6).strip() or "not out",
                }
                batting_list.append(batting_data)

        return batting_list

    def _extract_bowling(self, text: str) -> List[Dict[str, Any]]:
        """Extract bowling statistics from text"""
        bowling_list = []
        
        lines = [line.strip() for line in text.split('\n') if line.strip()]
        
        # Find bowling section
        bowling_start = -1
        for i, line in enumerate(lines):
            if re.search(r'(?:bowler|bowling)', line, re.IGNORECASE):
                bowling_start = i + 1
                break

        if bowling_start == -1:
            return bowling_list

        # Parse individual bowling rows
        # Pattern: Bowler Overs Maidens Runs Wickets Economy
        # Example: "Monika Chill 1.0 0 12 4 12.00"
        for line in lines[bowling_start:]:
            # Stop at next section
            if re.search(r'(?:extras|total|fall of wickets)', line, re.IGNORECASE):
                break

            bowling_match = re.match(
                r'^([A-Za-z0-9\s\(\)]+?)\s+([\d.]+)\s+(\d+)\s+(\d+)\s+(\d+)\s*(?:[\d.]+)?$',
                line
            )

            if bowling_match:
                bowler_name = bowling_match.group(1).strip()
                
                # Skip headers
                if bowler_name.lower() in ['no', 'bowler', 'overs', 'runs', 'wickets']:
                    continue

                bowling_data = {
                    "bowler": bowler_name,
                    "overs": float(bowling_match.group(2)),
                    "maidens": int(bowling_match.group(3)),
                    "runs": int(bowling_match.group(4)),
                    "wickets": int(bowling_match.group(5)),
                }
                bowling_list.append(bowling_data)

        return bowling_list

    def _calc_confidence(self, result: Dict) -> float:
        """Calculate parsing confidence score (0-1)"""
        score = 0.0
        max_score = 0.0

        # Match info confidence
        if result['match_info'].get('match'):
            score += 0.2
        max_score += 0.2

        # Batting data confidence
        if len(result['batting']) >= 11:
            score += 0.4
        elif len(result['batting']) >= 5:
            score += 0.2
        max_score += 0.4

        # Bowling data confidence
        if len(result['bowling']) >= 5:
            score += 0.4
        elif len(result['bowling']) >= 3:
            score += 0.2
        max_score += 0.4

        return round(score / max_score if max_score > 0 else 0, 2)

    def merge_innings(self, inning1: Dict, inning2: Dict) -> Dict:
        """Merge two innings into single scorecard"""
        return {
            "inning_1": inning1,
            "inning_2": inning2,
            "result": self._determine_result(inning1, inning2),
        }

    def _determine_result(self, inn1: Dict, inn2: Dict) -> str:
        """Determine match result"""
        inn1_runs = sum(b.get('runs', 0) for b in inn1.get('batting', []))
        inn2_runs = sum(b.get('runs', 0) for b in inn2.get('batting', []))
        
        if inn1_runs > inn2_runs:
            return f"Team 1 won by {inn1_runs - inn2_runs} runs"
        elif inn2_runs > inn1_runs:
            return f"Team 2 won by {inn2_runs - inn1_runs} runs"
        else:
            return "Match tied"

File: backend/scorecard-parser.js
// Copyright (c) 2026 RAJ.Y — All rights reserved.
// CH_Eleven Fantasy Cricket Platform
// https://github.com/RAJ-Y/ch-eleven
// Scorecard Parser - Extracts player stats from CSV/JSON scorecard data
// Supports both inning formats and bulk uploads

function parseScorecard(data, format = 'json') {
  if (format === 'csv') {
    return parseCSV(data);
  } else if (format === 'json') {
    return parseJSON(data);
  }
  throw new Error('Unsupported format. Use csv or json.');
}

function parseJSON(jsonStr) {
  try {
    const data = typeof jsonStr === 'string' ? JSON.parse(jsonStr) : jsonStr;
    
    const result = {
      matchInfo: data.match_info || {},
      batting: [],
      bowling: [],
      inning: data.inning || 1,
    };

    // Parse batting stats
    if (data.batting && Array.isArray(data.batting)) {
      result.batting = data.batting.map(b => ({
        player_name: b.batsman || b.name || '',
        runs: parseInt(b.runs || b.r || 0),
        balls_faced: parseInt(b.balls || b.b || 0),
        fours: parseInt(b.fours || b['4s'] || 0),
        sixes: parseInt(b.sixes || b['6s'] || 0),
        minutes: parseInt(b.minutes || b.m || 0),
        status: b.status || b.dismissal || '',
        team: b.team || '',
      }));
    }

    // Parse bowling stats
    if (data.bowling && Array.isArray(data.bowling)) {
      result.bowling = data.bowling.map(b => ({
        player_name: b.bowler || b.name || '',
        overs_bowled: parseFloat(b.overs || b.o || 0),
        maidens: parseInt(b.maidens || b.m || 0),
        runs_conceded: parseInt(b.runs || b.r || 0),
        wickets: parseInt(b.wickets || b.w || 0),
        team: b.team || '',
      }));
    }

    return result;
  } catch (e) {
    throw new Error(`JSON parse error: ${e.message}`);
  }
}

function parseCSV(csvStr) {
  const lines = csvStr.trim().split('\n');
  if (lines.length < 2) throw new Error('CSV must have headers and at least one data row');

  const result = {
    matchInfo: {},
    batting: [],
    bowling: [],
    inning: 1,
  };

  let section = null;
  
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    
    if (!line) continue;
    
    // Section headers
    if (line.toLowerCase().includes('[batting]')) {
      section = 'batting';
      i++; // skip header row
      continue;
    }
    if (line.toLowerCase().includes('[bowling]')) {
      section = 'bowling';
      i++; // skip header row
      continue;
    }

    const parts = line.split(',').map(p => p.trim());

    if (section === 'batting' && parts.length >= 3) {
      result.batting.push({
        player_name: parts[0],
        runs: parseInt(parts[1]) || 0,
        balls_faced: parseInt(parts[2]) || 0,
        fours: parseInt(parts[3]) || 0,
        sixes: parseInt(parts[4]) || 0,
        minutes: parseInt(parts[5]) || 0,
        status: parts[6] || '',
        team: parts[7] || '',
      });
    }

    if (section === 'bowling' && parts.length >= 5) {
      result.bowling.push({
        player_name: parts[0],
        overs_bowled: parseFloat(parts[1]) || 0,
        maidens: parseInt(parts[2]) || 0,
        runs_conceded: parseInt(parts[3]) || 0,
        wickets: parseInt(parts[4]) || 0,
        team: parts[5] || '',
      });
    }
  }

  return result;
}

// Match player name from scorecard to DB player
function matchPlayer(playerName, dbPlayers) {
  if (!playerName) return null;
  
  const normalized = playerName.toLowerCase().trim();
  
  // Exact match
  let match = dbPlayers.find(p => p.name.toLowerCase() === normalized);
  if (match) return match;
  
  // Partial match (first/last name)
  match = dbPlayers.find(p => 
    p.name.toLowerCase().includes(normalized) || 
    normalized.includes(p.name.toLowerCase())
  );
  
  return match || null;
}

// Build update payload for a player from scorecard data
function buildPlayerUpdate(scorecardPlayer, dbPlayer, type) {
  if (!dbPlayer) return null;
  
  const update = {
    id: dbPlayer.id,
    ...dbPlayer,
  };

  if (type === 'batting') {
    update.runs = (update.runs || 0) + scorecardPlayer.runs;
    update.balls_faced = (update.balls_faced || 0) + scorecardPlayer.balls_faced;
    update.fours = (update.fours || 0) + scorecardPlayer.fours;
    update.sixes = (update.sixes || 0) + scorecardPlayer.sixes;
    if (scorecardPlayer.status && scorecardPlayer.status.toLowerCase().includes('duck')) {
      update.duck = true;
    }
  }

  if (type === 'bowling') {
    update.wickets = (update.wickets || 0) + scorecardPlayer.wickets;
    update.overs_bowled = (update.overs_bowled || 0) + scorecardPlayer.overs_bowled;
    update.runs_conceded = (update.runs_conceded || 0) + scorecardPlayer.runs_conceded;
    update.maidens = (update.maidens || 0) + scorecardPlayer.maidens;
  }

  return update;
}

module.exports = {
  parseScorecard,
  parseJSON,
  parseCSV,
  matchPlayer,
  buildPlayerUpdate,
};
